src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -678,55 +678,201 @@ private static IEnumerable<TSource> TakeWhileIterator<TSource>(IEnumerable<TSour
 
         public static IEnumerable<TSource> Skip<TSource>(this IEnumerable<TSource> source, int count)
         {
-            if (source == null) throw Error.ArgumentNull("source");
-            return SkipIterator<TSource>(source, count);
+            Error.CheckSourceNotNull(source);
+            return new SkipIterator<TSource>(source, count);
         }
 
-        private static IEnumerable<TSource> SkipIterator<TSource>(IEnumerable<TSource> source, int count)
+        private class SkipIterator<TSource> : IEnumerable<TSource>
         {
-            using (IEnumerator<TSource> e = source.GetEnumerator())
+            // Just in case somebody does something weird with the ordering of enumerator
+            // disposal, this wraps an enumerator that has been used, but not disposed,
+            // and disposes it when this is.
+            private class DeadEnumeratorWrapper : IEnumerator<TSource>
             {
-                while (count > 0 && e.MoveNext()) count--;
-                if (count <= 0)
+                private readonly IEnumerator<TSource> _corpse;
+                public DeadEnumeratorWrapper(IEnumerator<TSource> corpse)
+                {
+                    _corpse = corpse;
+                }
+                public TSource Current
+                {
+                    get { return default(TSource); }
+                }
+                object IEnumerator.Current
+                {
+                    get { return default(TSource); }
+                }
+                public bool MoveNext()
+                {
+                    return false;
+                }
+                public void Dispose()
+                {
+                    _corpse.Dispose();
+                }
+                public void Reset()
+                {
+                    throw new NotSupportedException();
+                }
+            }
+            private readonly IEnumerable<TSource> _source;
+            private readonly int _count;
+            public SkipIterator(IEnumerable<TSource> source, int count)
+            {
+                _source = source;
+                _count = count > 0 ? count : 0;
+            }
+            public IEnumerator<TSource> GetEnumerator()
+            {
+                IEnumerator<TSource> sourceEnumerator = _source.GetEnumerator();
+                try
+                {
+                    for (int count = _count; count != 0; --count)
+                    {
+                        // It should be harmless to have the recipient call MoveNext again.
+                        // but we guard against strange behaviour from shortcuts in the enumerator
+                        // It should also be harmless to dispose the enumerator here if we've
+                        // exhausted it, but we guard against strange behaviour in the caller.
+                        if (!sourceEnumerator.MoveNext()) return new DeadEnumeratorWrapper(sourceEnumerator);
```

Comment:
The change seems observable and will break code relying on delayed execution of GetEnumerator and MoveNext. It is not hard to imagine such code to exist. 
However inconvenient, the former order of sideeffects needs to be preserved.